include/seqan3/utility/views/join_with.hpp

```diff
@@ -6,26 +6,525 @@
+template <class R, class P>
+concept compatible_joinable_ranges =
+    std::common_with<std::ranges::range_value_t<R>, std::ranges::range_value_t<P>>
+    && std::common_reference_with<std::ranges::range_reference_t<R>, std::ranges::range_reference_t<P>>
+    && std::common_reference_with<std::ranges::range_rvalue_reference_t<R>, std::ranges::range_rvalue_reference_t<P>>;
+
+template <class R>
+concept bidirectional_common = std::ranges::bidirectional_range<R> && std::ranges::common_range<R>;
+
+// A helper to decide whether the join_with view should have a non_propagating_cache member.
+template <typename InnerRng>
+struct cache_helper
+{};
+
+// If the inner range is a reference, we do not need a cache.
+template <typename InnerRng>
+    requires std::is_reference_v<InnerRng>
+struct cache_helper<InnerRng>
+{};
+
+// If the inner range is not a reference, we need a cache.
+template <typename InnerRng>
+    requires (!std::is_reference_v<InnerRng>)
+struct cache_helper<InnerRng>
+{
+    non_propagating_cache<std::remove_cv_t<InnerRng>> inner_;
+};
+
+} // namespace seqan3::detail::join_with
+
+namespace seqan3::detail
+{
+
+template <std::ranges::input_range V, std::ranges::forward_range Pattern>
+    requires std::ranges::view<V> && std::ranges::input_range<std::ranges::range_reference_t<V>>
+              && std::ranges::view<Pattern>
+              && join_with::compatible_joinable_ranges<std::ranges::range_reference_t<V>, Pattern>
+class join_with_view :
+    public std::ranges::view_interface<join_with_view<V, Pattern>>,
+    private join_with::cache_helper<std::ranges::range_reference_t<V>>
+{
+    using InnerRng = std::ranges::range_reference_t<V>;
+
+    V base_ = V();
+
+    Pattern pattern_ = Pattern();
+
+    template <bool Const>
+    struct iterator;
+
+    template <bool Const>
+    struct sentinel;
+
+public:
+    join_with_view()
+        requires std::default_initializable<V> && std::default_initializable<Pattern>
+    = default;
+
+    constexpr join_with_view(V base, Pattern pattern) : base_(std::move(base)), pattern_(std::move(pattern))
+    {}
+
+    template <std::ranges::input_range R>
+        requires std::constructible_from<V, std::ranges::views::all_t<R>>
+                  && std::constructible_from<Pattern, std::ranges::single_view<std::ranges::range_value_t<InnerRng>>>
+    constexpr join_with_view(R && r, std::ranges::range_value_t<InnerRng> e) :
+        base_(std::ranges::views::all(std::forward<R>(r))),
+        pattern_(std::ranges::views::single(std::move(e)))
+    {}
+
+    constexpr V base() const &
+        requires std::copy_constructible<V>
+    {
+        return base_;
+    }
+
+    constexpr V base() &&
+    {
+        return std::move(base_);
+    }
+
+    constexpr auto begin()
+    {
+        constexpr bool use_const =
+            view_helper::simple_view<V> && std::is_reference_v<InnerRng> && view_helper::simple_view<Pattern>;
+        return iterator<use_const>{*this, std::ranges::begin(base_)};
```

Comment:
why do we enforce const? (not sure what the above `use_const` means semantically)